ZEP000-wf-stream-processing.md

```diff
@@ -0,0 +1,668 @@
+* apply the business logic  
+* write updates to the state
+* write events and (optionally) follow-up commands to continue the processing
+* apply side-effects (e.g. send a response)
+
+The following diagram shows the command processing loop.
+
+![command-processing](ZEP-wf-stream-processing/event-sourcing-processing.png)
+
+The **events** are processed only on replay and rebuild the state. The replay of an event includes:
+* write updates to the state
+
+The following diagram shows the event replay.
+
+![event-replay](ZEP-wf-stream-processing/event-sourcing-replay.png)
+
+In order to rebuild the state with the events, it is necessary that all state changes from the command processing can be applied with the events only.
+So, the command processing may only update the state when it also writes an event that reflects the state changes.
+
+When the replay of the events is completed, the resulting state is exactly the same as when processing stopped the last time.
+Therefore, the processing of the commands can start again now.
+
+## The boundary between replay and processing
+
+On restart, it replays all events of the stream, but ignores events that have a source record position that is less than or equal to the last processed command of the snapshot. The state changes of these events are already included in the snapshot, because the commands are processed before the snapshot was created.
+ 
+When the replay is done, then the processing starts on the first command that was not processed before. A command is only processed if the stream contains a record (a command, an event, or a rejection) with a source record position referencing the command.
+
+![replay-and-processing](ZEP-wf-stream-processing/event-sourcing-replay-and-processing.png) 
+
+## Updatable engine
+
+Let's look at an example scenario in which we update the engine with business
+logic changes that leads to failure with the current stream processing, and how
+this proposal resolves the problem.
+
+Consider a compacted log stream `L` with records, where each record is numbered
+by its position in the log stream:
+
+10. `CreateWorkflowInstanceIntent:Create{workflow: 1}`
+11. `CreateWorkflowInstanceIntent:Created{workflow: 1, instance: 2}`
+12. `Process:Element_Activating{instance: 2}`
+
+And consider a snapshot `S1` containing the state after processing all records up
+to (but excluding) the record at position 10. This state contains a deployed
+workflow with key `1` and a Process BPMN element.
+
+**In the current engine**, the state of `S1` is copied and reprocessing starts
+at `L10` (i.e. the record in the log stream with position 10).
+
+First, during the reprocessing of `L10` a new key `2` is generated for the new
+workflow instance and this workflow instance is stored in the state. Next, the
+reprocessing of `L11` is skipped, because there is no processor for this record.
+Finally, processing starts again at `L12`, during which the Process BPMN element
```

Comment:
I don't see what this adds, can you explain why you would prefer this change?

Reply:
Maybe I got confused here. 

The confusion is about processing vs. reprocessing (especially in the context of the replay vs. processing). 

So my question is: Is L12 the last step of reprocessing or the first step of (the new) processing?

Reply:
Sorry for my late answer. Now that I read this again, I notice that I've left out valuable information. I forgot to mention that L12 has not yet been processed before. So the complete information is: The snapshot contains everything before L10, L10 has already been processed (producing L11 and L12) and it doesn't really matter whether or not L11 was already processed since no processor exists for it.

With that in mind, reprocessing is done on L10 and L11 and processing starts again at L12.
The same situation is true for the proposed processing/replay. L10 was already processed, so L11 should be replayed. L12 is then next to process after the replay finishes.